File: run_onnx/run_onnx.py
import migraphx
import sys
import numpy as np

def run_model(model, target):
    print("Running {} ...".format(target))

    model.compile(migraphx.get_target(target))
    params = {}
    args = []
    for key, value in model.get_parameter_shapes().items():
        print("Parameter {} -> {}".format(key, value))
        hash_val = hash(key) % (2 ** 32 - 1)
        np.random.seed(hash_val)
        args.append(np.random.randn(value.elements()).astype(np.single).reshape(value.lens()))
        params[key] = migraphx.argument(args[-1])

        result = np.array(model.run(params))

    print("")
    print("Result = \n{}".format(result))
    print("")

    return result

def main():
    if len(sys.argv) != 3:
        print("Usage: python {} onnx cpu/gpu/both".format(sys.argv[0]))
        exit()

    model = migraphx.parse_onnx(sys.argv[1])
    target = sys.argv[2]

    if target == "both":
        model_clone = model.clone()
        gpu_res = run_model(model_clone, "cpu")
        cpu_res = run_model(model, "gpu")
    elif target == "gpu":
        run_model(model, "gpu")
    else:
        run_model(model, "cpu")

if __name__ == "__main__":
    main()

File: run_onnx/run_onnx.cpp
#include <iostream>
#include <vector>
#include <string>
#include <migraphx/literal.hpp>
#include <migraphx/operators.hpp>
#include <migraphx/program.hpp>
#include <migraphx/instruction.hpp>
#include <migraphx/onnx.hpp>
#include "utilities.hpp"

migraphx::program load_onnx_file(std::string file_name) {
    auto prog = migraphx::parse_onnx(file_name);
    return prog;
}

int main(int argc, char **argv) {
    if (argc != 3) {
        std::cout << "Usage: " << argv[0] << " onnx_file cpu/gpu/both" << std::endl;
        return 0;
    }

    auto prog = load_onnx_file(argv[1]);
    int run_flag = 0;
    std::string str_flag(argv[2]);
    if (str_flag == "gpu")
    {
        run_flag = 1;
    }
    else if (str_flag == "both")
    {
        run_flag = 2;
    }

    std::vector<float> cpu_res, gpu_res;
    
    bool ret = true;
    if (run_flag == 0)
    {
        run_cpu(prog, cpu_res);
    }
    else if (run_flag == 1)
    {
        run_gpu(prog, gpu_res);
    }
    else 
    {
        run_cpu(prog, cpu_res);
        run_gpu(prog, gpu_res);
        ret = compare_results(cpu_res, gpu_res);
    }

    std::cout << (ret ? "PASSED!" : "FAILED") << std::endl;
    return 0;
}

